### ISSUE TITLE
unsigned conversion regression

### ISSUE THREAD

**Issue by mratsim** [nim-lang/Nim#12554]
The following used to give 44 on 0.20.2.

It now raises an error range error on 1.0.

```Nim
let x = 300'u
echo byte(x)
```

This broke our upcoming networking code: https://github.com/status-im/nim-libp2p/commit/541f0f2a41b0c47f41e63bd53fe8078e005629e2. In general, I don't think people expect range checking on unsigned integers, especially on converting to raw byte.

This may have a large impact on low-level code for example: graphics, crypto, networking, emulation ...

**Comment by kraptor** [comment]
For low-level code, what I always do is using proper casting to avoid this:

```
let x = 300'u
echo cast[byte](x)
```

**Comment by Araq** [comment]
Here is what the spec says about it:

```
`Automatic type conversion`:idx: is performed in expressions where different
kinds of integer types are used: the smaller type is converted to the larger.

A `narrowing type conversion`:idx: converts a larger to a smaller type (for
example ``int32 -> int16``. A `widening type conversion`:idx: converts a
smaller type to a larger type (for example ``int16 -> int32``). In Nim only
widening type conversions are *implicit*:

Type conversions
----------------
Syntactically a `type conversion` is like a procedure call, but a
type name replaces the procedure name. A type conversion is always
safe in the sense that a failure to convert a type to another
results in an exception (if it cannot be determined statically).

```

So it now behaves as the spec requires and previously it didn't. But I agree the wording is no explicit about unsigned types and worse there was no changelog entry for this change.

**Comment by Araq** [comment]
> This may have a large impact on low-level code for example: graphics, crypto, networking, emulation ...

We will add a changelog entry and we might add a switch to turn it off again, but looking at it without the history, it seems to be an improvement, use bit masking or ``cast`` to convert to bytes. Nim got more consistent in its behaviour.

**Comment by dryajov** [comment]
I generally agree with the behavior - it is safer,  but the problem is that it fails at runtime, it should be caught at compile time and `warn` at the very least, tho failing compilation is much more preferable.

**Comment by Araq** [comment]
Alright, good. Let's make this create a new warning.

**Comment by dryajov** [comment]
Dope!

**Comment by dryajov** [comment]
How about:

`Warning: narrowing type conversion are [deprecated/not allowed]`

**Comment by krux02** [comment]
What if this runtime check is exactly what I need?

**Comment by dryajov** [comment]
You're right, that one is caught, this one isn't:

```nim
let val: uint = 300
let myByte: byte = byte(val)
```

**Comment by Araq** [comment]
Then you turn off this warning? or rewrite your code to use ``cast`` or bitmasking.

**Comment by mratsim** [comment]
unsigned is not supposed to have overflow checks, when a number goes over the range it should roll over, including when narrowing the type.

Assuming we want to add all of that on the Nim unsigned integer, I would need new `memory int` types that behaves like the current unsigned int for the purposes of implementing VM and cryptography. Just like pointers and ptr UncheckedArray they would mean "I know what I'm doing".

Otherwise unsigned would be the first Nim feature with no low-level escape hatch.

**Comment by mratsim** [comment]
I was under the impression that casting was for reinterpreting the raw memory i.e. it would only be valid between types of the exact same size.

The old behaviour of byte was in 0.20.2 and before was the one I was describing.
The current behaviour is a regression introduce in 1.0 that has the potential of breaking all the past networking, graphics, VM, emulators, crypto and probably kernel in Nim projects codes written.

What is even worse is that it's a change that will not be caught at compile-time but at runtime.

And the ugly is that someone who wants to do not use this behavior has no recourse. I don't have alternative types that behave like a machine integer anymore. This is not expected from a system programming language, you need an escape hatch from those.

Lastly, in many cases and especially for cryptography, you might want not to throw any exceptions at all on certain part of the codebase. Now as soon as you use int or uint, you have an OverflowError tag that will pollute the effect system which makes one of the most promising usage of the effect system useless.

--------------

If you want the current 1.0 behavior you can use:
```Nim
proc safe_byte(x: int): byte =
  if x > high(int8):
    raise newError(RangeError, "Not in the 0 .. 255 range: " & $x)
  byte(x)
```

----------------

I want to emphasize that the issue is not just byte, it's all unsigned integer types as well.

### Unsigned -> unsigned

`Unsigned -> unsigned` conversion should be runtime check free

Works in 0.20.2 // Fails in 1.0.x
```Nim
let x = 1'u64 shl 32 + 1
echo uint32(x)
```

### Signed -> unsigned conversion

`Signed -> unsigned` should have range check because signed integers are checked.

This was the case also in the past
```Nim
let x = -1'i64
echo uint32(x)
```

### Signed -> signed conversion

`Signed -> signed` should have range check because signed integers are checked.

This was not the case in 0.20.2 and has been fixed in 1.0 (and introduced the current regression)
```Nim
let x = 1'u64 shl 32 + 1
echo int32(x)
```

**Comment by Araq** [comment]
I'm changing my mind about this issue and agree with @mratsim more and more.

**Comment by dryajov** [comment]
I'm on the same boat, after some cursory literature review, e.g. https://en.wikipedia.org/wiki/Integer_overflow (citation 1) - `a computation involving unsigned operands can never overflow`, as per C11 standard, I tend to agree with this being at the very least surprising behavior. Nim is ofcourse free to make its own choices in this regard, but this seems to be a well established practice.

My suggestion with regards to the warning where mostly to enable early detection of issues that this change introduced, other than that I now agree with @mratsim assessment.

**Comment by krux02** [comment]
I don't think that conversions to unsigned integers were not checked in the past. They were half ass range checked with special cases where range checking could not be implemented. Finally we have the int128 type in the compiler and we could implement all range checks correctly. `uint64` was such an exception type that was left out for range checks.

My suggestion is that you at least write an RFC for the behavior that you want, because you want to change the spec. 

@dryajov we are just talking about conversations into unsigned integer types. Not about computations within unsigned types. These already have the expected mod2 behavior.

**Comment by krux02** [comment]
Btw, I really think we should not touch this thing anymore. It was a big pain in the ass to get it all working correctly to the spec. It all works according to the spec. Using `cast[byet](x)` if you don't want a really dosn't seem to be a big deal. and btw, I made sure that casting between integer types of different sizes works correctly on the VM as well.

**Comment by dryajov** [comment]
> @dryajov we are just talking about conversations into unsigned integer types. Not about computations within unsigned types. These already have the expected mod2 behavior.

Then I would say that this is even less consistent, if computation already behaves in this way I would expect conversion to work the same.

**Comment by dryajov** [comment]
> Btw, I really think we should not touch this thing anymore. It was a big pain in the ass to get it all working correctly to the spec. It all works according to the spec. Using cast[byet](x) if you don't want a really dosn't seem to be a big deal. and btw, I made sure that casting between integer types of different sizes works correctly on the VM as well.

If we don't do something about it, existing code is going to trip on this pretty badly. At the very least this would require _everybody_ to audit their code or expect random crashes, I'm not sure this is the best experience for end users.

To avoid this we either a) revert to the previous behavior and allow conversion to behave just like computation does b) at the very least introduce a warning that allows people to find and fix their code.

**Comment by krux02** [comment]
@dryajov you want the old behavior where `var x: uint64 = -1` is fine, but `var y: uint8 = 300` is not?

**Comment by dryajov** [comment]
@krux02 I would go with what @mratsim has suggested so far, I.e. making sure that it is at least consistent with 0.20.2.

On a broader note, maybe arithmetic types need some revising, but I'm certainly not an expert on Nim nor type systems to have a super strong opinion. At this point tho, anything that touches this would probably be a breaking change and won't come before the next major release.

Btw, thanks for the awesome job on getting to 1.0. Nim is truly a special language and you've all done an amazing job!

**Comment by Araq** [comment]
> My suggestion is that you at least write an RFC for the behavior that you want, because you want to change the spec.

Correct, the bugfixes requires a spec update. I'm willing to do that because people have to come to rely on the old behaviour, so the spec should reflect the behavior of 0.20.

**Comment by mratsim** [comment]
> @dryajov you want the old behavior where `var x: uint64 = -1` is fine, but `var y: uint8 = 300` is not?
> 
> Just a reminder `let x = byte(300)` was always illegal.

This is about conversion between unsigned types. Your second example is using signed int.

This used to work
```Nim
let x = 300'u64
let y = byte(x)
```

Note that the fact that this doesn't `let x = byte(300'u64)` is due to the VM representing everything as an int internally for constant folding.

**Comment by krux02** [comment]
``let x = byte(300'u64)`` isn't computed in the VM. It is computed in semfold.nim.

**Comment by arnetheduck** [comment]
I would generally expect all conversions to unsigned not to raise - that's why I use unsigned in the first place, to get wrapping behaviour and I don't see why the function that converts to unsigned should be any different. the mirror operation, `uint->int` should raise however - `int` is a checked type

By this logic, `byte(-1)` is a totally permissible expression that resolves to the byte `0xff` both at runtime and at compile time (on a two's complement machine).

generally, I'd also think that `byte(300'u64)` _should_ be allowed - why should runtime and compile time have different behaviour? it is useful when wanting to find the bit pattern of some large number without calculating it in your head, and it's quite odd that `var x = 300'u64; byte(x)` would be handled differently.

**Comment by zah** [comment]
There is a way to follow the spec and to address @mratsim's concerns at the same time. 

The narrowing convention can be modelled to have a pre-condition that the value must be in range. When the compiler is not sure that this pre-condition is satisfied, it must insert a range check before the narrowing, but if the control flow demonstrates that the pre-condition is satisfied, the range check can be omitted.

How does the compiler know that the pre-condition is satisfied? Well, this is exactly the type of [data-flow analysis](https://en.wikipedia.org/wiki/Data-flow_analysis) that will become [possible](http://www.it.uu.se/research/group/astra/CPmeetsCAV/slides/qadeer.pdf) with the [planned integration](https://forum.nim-lang.org/t/5289) of Z3. It boils down to noticing that expressions such as `intVal and 0xff` are within a certain range and the same is also ensured by branches in the code such as `if value < 256: ...`.

This will allow us to have a simple non-casting template looking like this:

```nim
template truncate*(val: SomeUnsigned, TargetType: typedesc): TargetType =
  TargetType(val and high(TargetType))
```

The bit-masking should be included in the [peephole optimisation](https://en.wikipedia.org/wiki/Peephole_optimization) patterns of the back-end compiler and it produces exactly the same code as the "machine truncation" `byte(int_val)`. Check out the compiler code here:

https://godbolt.org/z/DD-uJc

Unguarded narrowing can become a warning for now and an error in the future. `truncate` can use a `cast` until the more sophisticated data-flow analysis is in place.

**Comment by Araq** [comment]
Here you go, https://github.com/nim-lang/RFCs/issues/175

**Comment by disruptek** [comment]
This code is currently failing to compile on Nim 1.0.2:

```nim
let t: uint = 3
assert 0 < t
```

This broke my little bump tool, if you can believe it.

**Comment by disruptek** [comment]
Another that broke golden:
```nim
let t: uint32 = 0
echo t xor 0o777 # fine
echo 0o777 xor t # nope
```

**Comment by Araq** [comment]
This one works for me fwiw:

```nim

let t: uint = 3
assert 0 < t
```

**Comment by disruptek** [comment]
Sorry; I should have been specific -- these are _now_ broken on **1.0.2** ~10/23 though they worked fine on **1.0.0** ~10/13.

**Comment by disruptek** [comment]
Never mind, I must have had 245a954b250b3e95f68c6bdadeba6ef1425442aa in my compiler, which means this **shouldn't** work in 1.0.  Duh.

**Comment by Araq** [comment]
Fixed, 1.0.4 will have it too.
